cql3/untyped_result_set.cc

```diff
@@ -47,58 +47,118 @@
 #include "result_set.hh"
 #include "transport/messages/result_message.hh"
 
-cql3::untyped_result_set_row::untyped_result_set_row(const map_t& data)
-    : _data(data)
+cql3::untyped_result_set_row::untyped_result_set_row(const index_map& index, const cql3::metadata& metadata, data_views data)
+    : _name_to_index(index)
+    , _metadata(metadata)
+    , _data(std::move(data))
 {}
 
-cql3::untyped_result_set_row::untyped_result_set_row(const std::vector<lw_shared_ptr<column_specification>>& columns, std::vector<bytes_opt> data)
-: _columns(columns)
-, _data([&columns, data = std::move(data)] () mutable {
-    map_t tmp;
-    std::transform(columns.begin(), columns.end(), data.begin(), std::inserter(tmp, tmp.end()), [](lw_shared_ptr<column_specification> c, bytes_opt& d) {
-       return std::make_pair<sstring, bytes_opt>(c->name->to_string(), std::move(d));
-    });
-    return tmp;
-}())
-{}
+size_t cql3::untyped_result_set_row::index(const std::string_view& name) const {
+    auto i = _name_to_index.find(name);
+    return i != _name_to_index.end() ? i->second : std::numeric_limits<size_t>::max();
+}
 
 bool cql3::untyped_result_set_row::has(std::string_view name) const {
-    auto i = _data.find(name);
-    return i != _data.end() && i->second;
+    auto i = index(name);
+    if (i < _data.size()) {
+        return !std::holds_alternative<std::monostate>(_data.at(i));
+    }
+    return false;
+}
+
+cql3::untyped_result_set_row::view_type cql3::untyped_result_set_row::get_view(std::string_view name) const {
+    return std::visit(make_visitor(
+        [](std::monostate) -> view_type { throw std::bad_variant_access(); },
+        [](const view_type& v) -> view_type { return v; },
+        [](const bytes& b) -> view_type { return view_type(b); }
+    ), _data.at(index(name)));
+}
+
+const std::vector<lw_shared_ptr<cql3::column_specification>>& cql3::untyped_result_set_row::get_columns() const {
+    return _metadata.get_names();
 }
 
 using cql_transport::messages::result_message;
 
-cql3::untyped_result_set::untyped_result_set(const cql3::result_set& rs) {
-    auto& cn = rs.get_metadata().get_names();
-    for (auto& r : rs.rows()) {
-        // r is const ref. TODO: make this more efficient by either wrapping result set
-        // or adding modifying accessors to it.
-        _rows.emplace_back(cn, r);
-    }
+cql3::untyped_result_set::index_map_ptr cql3::untyped_result_set::make_index(const cql3::metadata& metadata) {
+    auto map = std::make_unique<untyped_result_set_row::index_map>();
+    auto& names = metadata.get_names();
+    size_t i = 0;
+    std::transform(names.begin(), names.end(), std::inserter(*map, map->end()), [&](const lw_shared_ptr<column_specification>& c) mutable {
+        return std::make_pair<std::string_view, size_t>(c->name->text(), i++);
+    });
+    return map;
 }
 
+struct cql3::untyped_result_set::visitor {
+    rows_type& rows;
+    const cql3::metadata& meta;
+    const index_map_ptr& index;
```

Comment:
Now moving `untyped_result_set` will cause a dangling reference, no?
I had something different in mind, i.e. `const index_map&`, which is a reference to whatever the `unique_ptr` points to (and that something doesn't move when we move the owning `untyped_result_set`)

Reply:
Ok nvm it's just the `visitor`, `untyped_result_set` doesn't move during the lifetime of the `visitor` so there's no problem.
But still, why `ptr&` instead of just `&`

Reply:
Smallest possible change. And avoid using long type reference.

Reply:
k